### DISCUSSION TITLE
C++ get_args()

### DISCUSSION THREAD

**Discussion by sasq64** [mruby/mruby#5431] [Show and tell]
I am working on making it easier to bind C++ code to MRuby.
Here is a part of it, a function for extracting arguments;

Typical usage;
```cpp
    auto [x,y, text] = get_args<float, float, std::string>(mrb);
```
(Automatically builds the argument string according to the types and creates a tuple).

```cpp

template <typename ARG>
constexpr void get_spec(size_t i, ARG, char* data);

template <>
constexpr void get_spec(size_t i, mrb_int, char* data)
{ data[i] = 'i'; }

template <>
constexpr void get_spec(size_t i, mrb_float, char* data)
{ data[i] = 'f'; }

template <>
inline void get_spec(size_t i, const char*, char* data)
{ data[i] = 'z'; }

template <typename T> struct ToMrb { using To = T; };
template <> struct ToMrb<std::string> { using To = const char*; };
template <> struct ToMrb<std::string const&> { using To = const char*; };
template <> struct ToMrb<float> { using To = mrb_float; };
template <> struct ToMrb<int> { using To = mrb_int; };

template <class... ARGS, size_t... A>
auto get_args(
    mrb_state* mrb, std::vector<mrb_value>* restv, std::index_sequence<A...>)
{
    std::tuple<typename ToMrb<ARGS>::To...> input;

    std::array<char, sizeof...(ARGS) + 2> spec;
    spec[sizeof...(ARGS)] = '*';
    spec[sizeof...(ARGS) + 1] = 0;
    (void)std::tuple{(get_spec(A, std::get<A>(input), spec.data()), 0)...};

    mrb_int n = 0;
    mrb_value* rest{};
    mrb_get_args(mrb, spec.data(), &std::get<A>(input)..., &rest, &n);
    if (n > 0) {
        for (int i = 0; i < n; i++) {
            restv->push_back(rest[i]);
        }
    }
    return input;
}

//! Get function args according to type list
//! ex
//!
//! auto [x,y, txt] = get_args<float, float, std::string>(mrb);
//! draw_text(x,y, txt);
//
template <class... ARGS>
auto get_args(mrb_state* mrb)
{
    return get_args<ARGS...>(
        mrb, nullptr, std::make_index_sequence<sizeof...(ARGS)>());
}

template <class... ARGS>
auto get_args(mrb_state* mrb, std::vector<mrb_value>& rest)
{
    return get_args<ARGS...>(
        mrb, &rest, std::make_index_sequence<sizeof...(ARGS)>());
}
```

**Comment by matz** [comment]
I know only old C++. I didn't know we can do such a thing. Impressive & magical.

**Comment by sasq64** [comment]
It's powerful, but sadly pretty unreadable.
I hope I can build on that though, and make something even more useful.
The best example is the _sol_ library used for binding to _Lua_ which is fantastic.

**Reply by suetanvil** [reply]
I started using C++ again around 4 years ago for a job and I'm surprised at how pleasant to use the modern language has become. The community seems to be moving toward a very rich template library with compiler features to support making them less error-prone. I find that the strategies I use with modern scripting languages (e.g. throw it in a Set and then sort it) now work with C++ as well.

**Comment by Asmod4n** [comment]
Ive written something which turns most c++ objects into mruby objects, its cool how c++ can do most of that at compile time so its virtually free at runtime.

https://github.com/Asmod4n/mruby-c-ext-helpers/blob/master/include/mruby/cpp_to_mrb_value.hpp

Ive also written a small wrapper so you can turn a c++ class into a mruby MRB_TT_DATA with automatic cleanup

https://github.com/Asmod4n/mruby-c-ext-helpers/blob/master/include/mruby/cpp_helpers.hpp

Oh and automatic conversion of any number type into the correct mruby number type.

https://github.com/Asmod4n/mruby-c-ext-helpers/blob/master/include/mruby/num_helpers.hpp (sadly have to use mruby/internal.h here since the bigint stuff is hidden there.

Happy hacking :3
